src/main/kotlin/org/elm/ide/intentions/MakeDecoderIntention.kt

```diff
@@ -0,0 +1,184 @@
+        ty.module == "Maybe" && ty.name == "Maybe" -> existing(ty) ?: "(${qual("nullable")} ${gen(ty.parameters[0])})"
+        ty.module == "Set" && ty.name == "Set" -> generateSet(ty)
+        ty.module == "Dict" && ty.name == "Dict" -> generateDict(ty)
+        else -> generateUnionFunc(ty)
+    }
+
+    private fun generateRecord(ty: TyRecord): String {
+        val alias = ty.alias ?: return "(Debug.todo \"Cannot decode records without aliases\")"
+        val cached = existing(ty) ?: funcsByTy[ty]?.name
+        if (cached != null) return cached
+
+        val qualifier = qualifierFor(alias.toRef())
+        val name = "decode${funcNames[alias.toRef()]}"
+        val body = buildString {
+            append("    ${qual("succeed")} $qualifier${alias.name}")
+            for ((field, fieldTy) in ty.fields) {
+                append("\n        |> required \"$field\" ${gen(fieldTy)}")
+            }
+        }
+        val func = GeneratedFunction(name = name, paramTy = ty, paramName = "", body = body, qualifier = qualifier)
+        funcsByTy[ty] = func
+        usedPipeline = true
+        return name
+    }
+
+    private fun generateUnionFunc(ty: TyUnion): String {
+        val cached = existing(ty) ?: funcsByTy[ty]?.name
+        if (cached != null) return cached
+
+        val decl: ElmTypeDeclaration? = ElmLookup.findFirstByNameAndModule(ty.name, ty.module, file)
+        val body = if (decl == null) {
+            "Debug.todo \"Can't generate decoder for ${ty.name}\""
+        } else {
+            unionsToExpose += ty.toRef()
+            val variants = decl.variantInference().value
+            if (variants.size == 1 && variants.values.first().size == 1) {
+                "    ${qual("map")} ${variants.keys.first()} ${gen(variants.values.first().first())}"
+            } else {
+                val branches = variants.entries.joinToString("\n\n                ") { (variant, params) ->
+                    val expr = when {
+                        params.isEmpty() -> "${qual("succeed")} ${qual(ty.module, variant)}"
+                        else -> "Debug.todo \"Cannot decode variant with params: $variant\""
+                    }
+                    """
+                    |"$variant" ->
+                    |                    $expr
+                    """.trimMargin()
+                }
+                """
+                |    let
+                |        get id =
+                |            case id of
+                |                $branches
+                |
+                |                _ ->
+                |                    ${qual("fail")} ("unknown value for ${ty.renderedText(false, false)}: " ++ id)
+                |    in
+                |    ${qual("string")} |> ${qual("andThen")} get
+                """.trimMargin()
+            }
+        }
+
+        val name = "decode${ty.alias?.let { funcNames[it.toRef()] } ?: ty.name}"
```

Comment:
Can you change it so that generated, nested functions are named `fooDecoder` instead of `decodeFoo`? This is the style that elm-spa-example and the [elm-guide](https://guide.elm-lang.org/effects/json.html) uses.